### ISSUE TITLE
DFU issues when Packet Receipt Notification (PRN) procedure is enabled 

### ISSUE THREAD

**Issue by plaes** [lulf/nrf-dfu#3]
There seem to be some edge cases when handling DFU protocol, so let's write the things down:

Firstly, two sets of logs with "previous" code based on watchful where `NrfDfuServiceEvent::PacketWrite` is using `packet_notify` instead of `notify_control`.

1. Working case (PRN disabled) and using `packet_notify`:
```
REQUEST Select { obj_type: Command }
RESPONSE DfuResponse { request: Select { obj_type: Command }, result: Success, body: Some(Select { offset: 0, crc: 0, max_size: 512 }) }
REQUEST SetReceiptNotification { target: 0 }
RESPONSE DfuResponse { request: SetReceiptNotification { target: 0 }, result: Success, body: None }
REQUEST Create { obj_type: Command, obj_size: 79 }
RESPONSE DfuResponse { request: Create { obj_type: Command, obj_size: 79 }, result: Success, body: None }
REQUEST Write { data: [..] }
RESPONSE DfuResponse { request: Write { data: [..] }, result: Success, body: Some(Crc { offset: 79, crc: 2913568115 }) }
REQUEST Crc
RESPONSE DfuResponse { request: Crc, result: Success, body: Some(Crc { offset: 79, crc: 2913568115 }) }
REQUEST Execute
RESPONSE DfuResponse { request: Execute, result: Success, body: None }
REQUEST Select { obj_type: Data }
RESPONSE DfuResponse { request: Select { obj_type: Data }, result: Success, body: Some(Select { offset: 0, crc: 0, max_size: 335872 }) }
REQUEST Create { obj_type: Data, obj_size: 61052 }
RESPONSE DfuResponse { request: Create { obj_type: Data, obj_size: 61052 }, result: Success, body: None }
REQUEST Write { data: [..] }
RESPONSE DfuResponse { request: Write { data: [..] }, result: Success, body: Some(Crc { offset: 120, crc: 1031809356 }) }
REQUEST Write { data: [..] }
RESPONSE DfuResponse { request: Write { data: [..] }, result: Success, body: Some(Crc { offset: 240, crc: 4174146515 }) }
```

2. Non-working case, where connection hangs after fifth packet (PRN is configured to 5 in NRF-Connect):
```
REQUEST Select { obj_type: Command }
RESPONSE DfuResponse { request: Select { obj_type: Command }, result: Success, body: Some(Select { offset: 0, crc: 0, max_size: 512 }) }
REQUEST SetReceiptNotification { target: 0 }
RESPONSE DfuResponse { request: SetReceiptNotification { target: 0 }, result: Success, body: None }
REQUEST Create { obj_type: Command, obj_size: 79 }
RESPONSE DfuResponse { request: Create { obj_type: Command, obj_size: 79 }, result: Success, body: None }
REQUEST Write { data: [..] }
RESPONSE DfuResponse { request: Write { data: [..] }, result: Success, body: Some(Crc { offset: 79, crc: 2913568115 }) }
REQUEST Crc
RESPONSE DfuResponse { request: Crc, result: Success, body: Some(Crc { offset: 79, crc: 2913568115 }) }
REQUEST Execute
RESPONSE DfuResponse { request: Execute, result: Success, body: None }
REQUEST SetReceiptNotification { target: 5 }
RESPONSE DfuResponse { request: SetReceiptNotification { target: 5 }, result: Success, body: None }
REQUEST Select { obj_type: Data }
RESPONSE DfuResponse { request: Select { obj_type: Data }, result: Success, body: Some(Select { offset: 0, crc: 0, max_size: 335872 }) }
REQUEST Create { obj_type: Data, obj_size: 61052 }
RESPONSE DfuResponse { request: Create { obj_type: Data, obj_size: 61052 }, result: Success, body: None }
REQUEST Write { data: [..] }
RESPONSE DfuResponse { request: Write { data: [..] }, result: Success, body: None }
REQUEST Write { data: [..] }
RESPONSE DfuResponse { request: Write { data: [..] }, result: Success, body: None }
REQUEST Write { data: [..] }
RESPONSE DfuResponse { request: Write { data: [..] }, result: Success, body: None }
REQUEST Write { data: [..] }
RESPONSE DfuResponse { request: Write { data: [..] }, result: Success, body: None }
REQUEST Write { data: [..] }
RESPONSE DfuResponse { request: Write { data: [..] }, result: Success, body: Some(Crc { offset: 600, crc: 2318566729 }) }
... hang ...
```

**Comment by plaes** [comment]
Second step, switching `NrfDfuServiceEvent::PacketWrite` to `notify_control` / `control_notify`:

NRF-Connect stops in both cases (PRN enabled/disabled) after initial response: 
```
on_data_length_update conn_handle=3 max_rx_octets=127 max_rx_time_us=1362 max_tx_octets=127 max_tx_time_us=1362
DFU REQUEST Select { obj_type: Command }
DFU RESPONSE DfuResponse { request: Select { obj_type: Command }, result: Success, body: Some(Select { offset: 0, crc: 0, max_size: 512 }) }
DFU REQUEST SetReceiptNotification { target: 0 }
DFU RESPONSE DfuResponse { request: SetReceiptNotification { target: 0 }, result: Success, body: None }
DFU REQUEST Create { obj_type: Command, obj_size: 79 }
DFU RESPONSE DfuResponse { request: Create { obj_type: Command, obj_size: 79 }, result: Success, body: None }
DFU REQUEST Write { data: [..] }
DFU RESPONSE DfuResponse { request: Write { data: [..] }, result: Success, body: Some(Crc { offset: 79, crc: 2913568115 }) }
DFU REQUEST Crc
DFU RESPONSE DfuResponse { request: Crc, result: Success, body: Some(Crc { offset: 79, crc: 2913568115 }) }
```
NRF-Connect:
```
[DFU] Command object sent (CRC = ADA98573)
[DFU] Writing to characteristic 8ec90001-f315-4f60-9fb8-838830daea50, value (0x): 03
[DFU] gatt.writeCharacteristic(8ec90001-f315-4f60-9fb8-838830daea50, value=0x03, WRITE_TYPE_DEFAULT)
[DFU] Notification received from 8ec90001-f315-4f60-9fb8-838830daea50, value (0x): 60-08-01-4F-00-00-00-73-85-A9-AD
[DFU] Data written to 8ec90001-f315-4f60-9fb8-838830daea50
[DFU] Invalid response received (response: 0x6008014F0000007385A9AD, expected: 0x6003..)
```

**Comment by plaes** [comment]
And now with the dfu-fix and `NrfDfuServiceEvent::PacketWrite` to `notify_control` / `control_notify`, initial connection fails after receiving `Execute` response (in both PRN enabled and disabled cases):
```
REQUEST Select { obj_type: Command }
RESPONSE DfuResponse { request: Select { obj_type: Command }, result: Success, body: Some(Select { offset: 0, crc: 0, max_size: 512 }) }
REQUEST SetReceiptNotification { target: 0 }
RESPONSE DfuResponse { request: SetReceiptNotification { target: 0 }, result: Success, body: None }
REQUEST Create { obj_type: Command, obj_size: 79 }
RESPONSE DfuResponse { request: Create { obj_type: Command, obj_size: 79 }, result: Success, body: None }
REQUEST Write { data: [..] }
RESPONSE DfuResponse { request: Crc, result: Success, body: Some(Crc { offset: 79, crc: 2913568115 }) } 
REQUEST Crc
RESPONSE DfuResponse { request: Crc, result: Success, body: Some(Crc { offset: 79, crc: 2913568115 }) } 
REQUEST Execute
RESPONSE DfuResponse { request: Execute, result: Success, body: None }
```
Logs from nrf-connect:
```
[DFU] Notification received from 8ec90001-f315-4f60-9fb8-838830daea50, value (0x): 60-03-01-4F-00-00-00-73-85-A9-AD
[DFU] Data written to 8ec90001-f315-4f60-9fb8-838830daea50
[DFU] Checksum received (Offset = 79, CRC = ADA98573)
[DFU] Writing to characteristic 8ec90001-f315-4f60-9fb8-838830daea50, value (0x): 04
[DFU] gatt.writeCharacteristic(8ec90001-f315-4f60-9fb8-838830daea50, value=0x04, WRITE_TYPE_DEFAULT)
[DFU] Notification received from 8ec90001-f315-4f60-9fb8-838830daea50, value (0x): 60-03-01-4F-00-00-00-73-85-A9-AD
[DFU] Data written to 8ec90001-f315-4f60-9fb8-838830daea50
[DFU] Invalid response received (response: 0x6003014F0000007385A9AD, expected: 0x6004..)
```

**Comment by plaes** [comment]
There's a fix in #4
